BUG: Fix argument invalid seek when startxref is zero

VictorCarlquist opened the pull request:
Hi!

When pdf file ends with the following lines below, the stream.seek(-11, 1) at line 1854 raises Exception Value (22) Argument invalid. 

    xref
    0 85
    0000000000 65535 f
    ....
    0000000000 00000 n
    trailer
    <<
    /Size 85
    /Root 84 0 R
    /Info 83 0 R

    startxref
    0
    %%EOF

So, when  `startxref` is zero, this commit expand the search to try to find the xref table in the last 2KB.

kurtmckee commented:
@VictorCarlquist, it sounds like this is a valuable addition to PyPDF2! Would you add a unit test to your commit (like an example file) that demonstrates the problem that you're fixing?

Thanks for this merge request!

VictorCarlquist commented:
Hi @kurtmckee, sorry for the big delay to answer.

With 2KB, the pyPDF keeps showing up the Exception with some pdf files, so we have raised the search to 8KB and it solved the problem. 

The 8KB seems fine to you? Could we keep this value?

MartinThoma commented on PyPDF2/pdf.py:
```diff
@@ -1694,7 +1694,7 @@ def _getObjectFromStream(self, indirectReference):
             try:
                 obj = readObject(streamData, self)
-            except utils.PdfStreamError as e:
+            except PdfStreamError as e:
```
```suggestion
            except utils.PdfStreamError as e:
```

MartinThoma commented:
@VictorCarlquist Would you mind creating a minimal example that shows the issue? Something like this:

```python
def test_get_images_raw():
    strict = True
    with_prev_0 = False
    pdf_data = (
        b"%%PDF-1.7\n"
        b"1 0 obj << /Count 1 /Kids [4 0 R] /Type /Pages >> endobj\n"
        b"2 0 obj << >> endobj\n"
        b"3 0 obj << >> endobj\n"
        b"4 0 obj << /Contents 3 0 R /CropBox [0.0 0.0 2550.0 3508.0]"
        b" /MediaBox [0.0 0.0 2550.0 3508.0] /Parent 1 0 R"
        b" /Resources << /Font << >> >>"
        b" /Rotate 0 /Type /Page >> endobj\n"
        b"5 0 obj << /Pages 1 0 R /Type /Catalog >> endobj\n"
        b"xref 1 5\n"
        b"%010d 00000 n\n"
        b"%010d 00000 n\n"
        b"%010d 00000 n\n"
        b"%010d 00000 n\n"
        b"%010d 00000 n\n"
        b"trailer << %s/Root 5 0 R /Size 6 >>\n"
        b"startxref %d\n"
        b"%%%%EOF"
    )
    pdf_data = pdf_data % (
        pdf_data.find(b"1 0 obj"),
        pdf_data.find(b"2 0 obj"),
        pdf_data.find(b"3 0 obj"),
        pdf_data.find(b"4 0 obj"),
        pdf_data.find(b"5 0 obj"),
        b"/Prev 0 " if with_prev_0 else b"",
        pdf_data.find(b"xref"),
    )
    pdf_stream = io.BytesIO(pdf_data)
    PdfFileReader(pdf_stream, strict=strict)  # just shows that nothing crashes
```

It should be a test that currently fails, but with your PR it doesn't.

VictorCarlquist commented:
Hi @MartinThoma, thank you for your feedback.

I added the tests and I changed the code a little.

Is it possible to install the `mock` package?

MartinThoma commented:
> Is it possible to install the mock package?

Sure - for testing on Py27:

1. Add it to `requirements/ci.in`
2. Run `pip-compile requirements/ci.in` to generate the `ci.txt` (pip-compile comes from pip-tools)

VictorCarlquist commented:
> > Is it possible to install the mock package?
> 
> Sure - for testing on Py27:
> 
>     1. Add it to `requirements/ci.in`
> 
>     2. Run `pip-compile requirements/ci.in` to generate the `ci.txt` (pip-compile comes from pip-tools)

I did it but it's keeping failing, seems that the CI does't get the new ci.txt file.

MartinThoma commented:
Damn, sorry, I forgot the hack we have for Python 2.7.

For Python 2.7 you need to add mock here: https://github.com/py-pdf/PyPDF2/blob/main/.github/workflows/github-ci.yaml#L45

VictorCarlquist commented:
> Damn, sorry, I forgot the hack we have for Python 2.7.
> 
> For Python 2.7 you need to add mock here: https://github.com/py-pdf/PyPDF2/blob/main/.github/workflows/github-ci.yaml#L45

Thanks Martin!

MartinThoma reviewed (changes requested):
PyPDF2 has changed quite a bit since you made this PR. Could you please check if it still is necessary?

We now have:

```
    def _find_startxref_pos(self, stream: StreamType) -> int:
        """Find startxref entry - the location of the xref table"""
        line = read_previous_line(stream)
        try:
            startxref = int(line)
        except ValueError:
            # 'startxref' may be on the same line as the location
            if not line.startswith(b"startxref"):
                raise PdfReadError("startxref not found")
            startxref = int(line[9:].strip())
            warnings.warn("startxref on same line as offset", PdfReadWarning)
        else:
            line = read_previous_line(stream)
            if line[:9] != b"startxref":
                raise PdfReadError("startxref not found")
        return startxref
```

In case you want to look for that part of the code: Search for

```
last_mb = stream.tell() - 1024 * 1024 + 1  # offset of last MB of stream
```
